packages/attestation-service/src/sms/index.ts

```diff
@@ -100,197 +120,300 @@ export function unsupportedRegionCodes() {
   return intersection(smsProviders.map((provider) => provider.unsupportedRegionCodes))
 }
 
-// Main entry point for sending SMS via preferred providers.
+function providersToCsv(providers: SmsProvider[]) {
+  return providers.map((provider) => provider.type).join(',')
+}
+
+// This list may be a strict subset of the valid providers (e.g. when forcing a provider for a test),
+// but cannot require providers not configured or unsupported for this country code.
+function getProvidersFor(attestation: AttestationModel, logger: Logger) {
+  const validProviders = smsProvidersFor(attestation.countryCode, attestation.phoneNumber, logger)
+  const attestationProviders = attestation.providers
+    .split(',')
+    .filter((t) => t != null && t !== '')
+    .map((name) => smsProvidersByType[name])
+
+  for (const p of attestationProviders) {
+    if (!validProviders.find((v) => p.type === v.type)) {
+      throw new Error(`Detected inconsistent provider configuration between instances`)
+    }
+  }
+  return attestationProviders
+}
 
+// Main entry point for sending SMS.
 export async function startSendSms(
+  key: AttestationKey,
   phoneNumber: E164Number,
   message: string,
-  finallyFailedCallback?: () => void,
-  finallyBelievedDeliveredCallback?: () => void
-): Promise<SmsProviderType> {
-  const countryCode = phoneUtil.getRegionCodeForNumber(phoneUtil.parse(phoneNumber))
-  if (!countryCode) {
-    throw new Error('Could not parse number')
-  }
+  logger: Logger,
+  sequelizeLogger: SequelizeLogger,
+  onlyUseProvider: string | null = null
+): Promise<AttestationModel> {
+  let shouldRetry = false
+  let attestation: AttestationModel | null = null
 
-  const providers = smsProvidersFor(countryCode, phoneNumber)
-  if (providers.length === 0) {
-    Counters.attestationRequestsUnableToServe.labels(countryCode).inc()
-    throw new Error(`No SMS providers available for ${countryCode}`)
+  const transaction = await sequelize!.transaction({ logging: sequelizeLogger })
+
+  try {
+    const parsedNumber = phoneUtil.parse(phoneNumber)
+    const numberType = parsedNumber ? phoneUtil.getNumberType(parsedNumber) : null
+    const countryCode = parsedNumber ? phoneUtil.getRegionCodeForNumber(parsedNumber) : null
+
+    let providers: SmsProvider[] = countryCode
+      ? smsProvidersFor(countryCode, phoneNumber, logger)
+      : []
```

Comment:
any reason not to just have the default provider list here? or are we certain that providers will fail to send if phoneUtil can not get the country code?

Reply:
in fact this is a failsafe. Parsing the request itself fails if the phone number is not E164 format.